driver/ni/digital_write.cpp

```diff
@@ -158,77 +178,103 @@ int ni::DigitalWriteSink::init(){
+        }
+    });
     return freighter::NIL;
 }
 
 
-freighter::Error ni::DigitalWriteSink::stop(){
-    if(!this->running.exchange(false)){
+freighter::Error ni::DigitalWriteSink::stop() {
+    if (!this->running.exchange(false)) {
         return freighter::NIL;
     }
-    if (this->checkNIError(ni::NiDAQmxInterface::StopTask(task_handle))){
-        LOG(ERROR) << "[NI Writer] failed while stopping writer for task " << this->writer_config.task_name;
+    if (this->checkNIError(ni::NiDAQmxInterface::StopTask(task_handle))) {
+        LOG(ERROR) << "[NI Writer] failed while stopping writer for task " << this->
+                writer_config.task_name;
         return freighter::Error(driver::CRITICAL_HARDWARE_ERROR);
     }
-    LOG(INFO) << "[NI Writer] successfully stopped writer for task " << this->writer_config.task_name;
+    LOG(INFO) << "[NI Writer] successfully stopped writer for task " << this->
+            writer_config.task_name;
+    ctx->setState({
+        .task = this->writer_config.task_key,
+        .variant = "success",
+        .details = {
+            {"running", false}
+        }
+    });
     return freighter::NIL;
 }
 
-freighter::Error ni::DigitalWriteSink::write(synnax::Frame frame){
-    char errBuff[2048] = {'\0'};
+freighter::Error ni::DigitalWriteSink::write(synnax::Frame frame) {
     int32 samplesWritten = 0;
     formatData(std::move(frame));
 
     // Write digital data
     if (this->checkNIError(ni::NiDAQmxInterface::WriteDigitalLines(this->task_handle,
-                                                                   1,                        // number of samples per channel
-                                                                   1,                        // auto start
-                                                                   10.0,                     // timeout
-                                                                   DAQmx_Val_GroupByChannel, // data layout
-                                                                   writeBuffer,              // data
-                                                                   &samplesWritten,          // samples written
-                                                                   NULL))){
-        LOG(ERROR) << "[NI Writer] failed while writing digital data for task " << this->writer_config.task_name;
-        return freighter::Error(driver::CRITICAL_HARDWARE_ERROR, "Error reading digital data");
+        1, // number of samples per channel
+        1, // auto start
+        10.0, // timeout
+        DAQmx_Val_GroupByChannel, // data layout
+        writeBuffer, // data
+        &samplesWritten, // samples written
+        NULL))) {
+        LOG(ERROR) << "[NI Writer] failed while writing digital data for task " << this
+                ->writer_config.task_name;
+        return freighter::Error(driver::CRITICAL_HARDWARE_ERROR,
+                                "Error reading digital data");
```

Comment:
Is there a way we can extract a better error message here?

Reply:
the more verbose error will get sent to task state through check ni error. I could provide that here too but it would just be the same message